### PR TITLE
feat(pip): implement PIP-23

### LINKED ISSUE #1386: PIP-23: Sign and verify message by public and private key account
## Description

The purpose of this proposal is to introduce a standardized method for signing and verifying messages using public and private key pairs associated with Pactus accounts.

https://pips.pactus.org/PIPs/pip-23

### PR DESCRIPTION
**Pull request by ZigBalthazar**
- Fixes #1386

### CODE REVIEW

**Review by kehiy — CHANGES_REQUESTED** [review]
Please run `make proto` and make sure it finishs proprtly.

File: util/crypto/crypto.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	if err != nil {
+		return false, err
+	}
+
+	pub, err := bls.PublicKeyFromString(pubStr)
+	if err != nil {
+		return false, err
+	}
+
+	if err := pub.Verify([]byte(msg), sig); err != nil {
```
Comment (Ja7ad) [inline util/crypto/crypto.go:25]: Function `VerifyMessage` just return true or false, not required to handle error.

File: util/crypto/crypto_test.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	msg := "pactus"
+	prvStr := "SECRET1PDRWTLP5PX0FAHDX39GXZJP7FKZFALML0D5U9TT9KVQHDUC99CMGQQJVK67"
+	sigStr := "923d67a8624cbb7972b29328e15ec76cc846076ccf00a9e94d991c677846f334ae4ba4551396fbcd6d1cab7593baf3b7"
+
+	sig, err := SignMessageWithPrivateKey(prvStr, msg)
+	assert.Nil(t, err)
+	assert.Equal(t, sig, sigStr)
+}
+
+func TestVerifyMessage(t *testing.T) {
```
Comment (Ja7ad) [inline util/crypto/crypto_test.go:19]: Please check invalid signature.

File: util/crypto/crypto_test.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSignMessageWithPrivateKey(t *testing.T) {
+	msg := "pactus"
+	prvStr := "SECRET1PDRWTLP5PX0FAHDX39GXZJP7FKZFALML0D5U9TT9KVQHDUC99CMGQQJVK67"
+	sigStr := "923d67a8624cbb7972b29328e15ec76cc846076ccf00a9e94d991c677846f334ae4ba4551396fbcd6d1cab7593baf3b7"
+
+	sig, err := SignMessageWithPrivateKey(prvStr, msg)
```
Comment (Ja7ad) [inline util/crypto/crypto_test.go:14]: Please check invalid private key

**Review by Ja7ad — CHANGES_REQUESTED** [review]
Please write utils service (`utils.proto`) for sign and verify message outside of wallet.

https://github.com/pactus-project/pactus/blob/5727751c38775e35a2a67a324c657289927a0a9a/www/grpc/buf/grpc-gateway.config.yaml#L4

```yaml
    - selector: pactus.Utils.SignMessage
      get: "/pactus/utils/sign_message"

    - selector: pactus.Utils.VerifyMessage
      get: "/pactus/utils/verify_message"
```

https://pips.pactus.org/PIPs/pip-23#specification

**Comment by ZigBalthazar** [comment]
> Please write util service for sign and verify message outside of wallet.
> 
> https://pips.pactus.org/PIPs/pip-23#specification

Hi @Ja7ad 
regarding to PIP-23 spec, the `SignMessageWithPrivateKey` and `VerifyMessage` should implement in util and the `SignMessage` should be in wallet service
is it right?

**Comment by ZigBalthazar** [comment]
Hi @Ja7ad ,
in my opinion we should not put the `sign message` in grpc cause of the sign message need private Key and if we transfer the private key on network could make some security vulnerability

**Comment by Ja7ad** [comment]
> Hi @Ja7ad ,
> in my opinion we should not put the `sign message` in grpc cause of the sign message need private Key and if we transfer the private key on network could make some security vulnerability 

You should put it, don't have vulnerability.

File: util/crypto/crypto.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package crypto
+
+import "github.com/pactus-project/pactus/crypto/bls"
+
+func SignMessageWithPrivateKey(prv, msg string) (string, error) {
```
Comment (b00f) [inline util/crypto/crypto.go:5]: Please don't define extra util when we have `crypto` package. Use crypto package.
Here we have a wrapped function.

File: wallet/wallet_test.go [hunk 5]
```diff
@@ -569,3 +570,21 @@ func TestTotalBalance(t *testing.T) {
 	assert.NoError(t, err)
 	assert.Equal(t, totalBalance, acc1.Balance()+acc3.Balance())
 }
+
+func TestSignMessage(t *testing.T) {
+	td := setup(t)
+	defer td.Close()
+
+	msg := "pactus"
+	sigStr := "923d67a8624cbb7972b29328e15ec76cc846076ccf00a9e94d991c677846f334ae4ba4551396fbcd6d1cab7593baf3b7"
```
Comment (b00f) [inline wallet/wallet_test.go:579]: Rename it to `expectedSig`

**Review by b00f — CHANGES_REQUESTED** [review]

File: www/grpc/proto/util.proto [hunk 6]
```diff
@@ -0,0 +1,43 @@
+syntax = "proto3";
+package pactus;
+
+option go_package = "github.com/pactus-project/pactus/www/grpc/pactus";
+option java_package = "pactus.util";
+
+// Util service defines various RPC methods for interacting with
+// Utils.
+service Util {
```
Comment (Ja7ad) [inline www/grpc/proto/util.proto:9]: Please change `Util` to `Utils`

File: www/grpc/buf/grpc-gateway.config.yaml [hunk 7]
```diff
@@ -87,3 +87,13 @@ http:
 
     - selector: pactus.Wallet.GetTotalBalance
       get: "/pactus/wallet/get_total_balance"
+
+    - selector: pactus.Wallet.SignMessage
+      get: "/pactus/wallet/sign_message"
+
+    # Util APIs
+    - selector: pactus.Util.SignMessageWithPrivateKey
+      get: "/pactus/Util/Sign_Message_With_PrivateKey"
```
Comment (Ja7ad) [inline www/grpc/buf/grpc-gateway.config.yaml:96]: Change to `Utils`

File: www/grpc/buf/grpc-gateway.config.yaml [hunk 8]
```diff
@@ -87,3 +87,13 @@ http:
...
+
+    - selector: pactus.Util.VerifyMessage
+      get: "/pactus/Util/VerifyMessage"
```
Comment (Ja7ad) [inline www/grpc/buf/grpc-gateway.config.yaml:99]: Change to `Utils`

**Review by Ja7ad — CHANGES_REQUESTED** [review]

File: wallet/wallet_test.go [hunk 9]
```diff
@@ -569,3 +570,21 @@ func TestTotalBalance(t *testing.T) {
...
 }
+
+func TestSignMessage(t *testing.T) {
+	td := setup(t)
+	defer td.Close()
+
+	msg := "pactus"
+	expectedSig := "923d67a8624cbb7972b29328e15ec76cc846076ccf00a9e94d991c677846f334ae4ba4551396fbcd6d1cab7593baf3b7"
+
+	prv, _ := bls.PrivateKeyFromString(
```
Comment (b00f) [inline wallet/wallet_test.go:581]: Make sure the error is nil:

```
prv, err := bls.PrivateKeyFromString("...")
require.NoError(t, err)

```

File: www/grpc/buf/grpc-gateway.config.yaml [hunk 10]
```diff
@@ -87,3 +87,13 @@ http:
...
+
+    - selector: pactus.Wallet.SignMessage
+      get: "/pactus/wallet/sign_message"
+
+    # Util APIs
+    - selector: pactus.Utils.SignMessageWithPrivateKey
+      get: "/pactus/Utils/Sign_Message_With_PrivateKey"
+
+    - selector: pactus.Utils.VerifyMessage
+      get: "/pactus/Utils/VerifyMessage"
```
Comment (b00f) [inline www/grpc/buf/grpc-gateway.config.yaml:99]: Add new line here

Comment (Ja7ad) [inline www/grpc/buf/grpc-gateway.config.yaml]: Please change

```
/pactus/Utils/verify_message
```

**Review by Ja7ad — CHANGES_REQUESTED** [review]
@ZigBalthazar please register utils service in json-rpc server.

**Review by b00f — CHANGES_REQUESTED** [review]

File: www/grpc/proto/utils.proto [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+  // VerifyMessage verify signature with public key and message
+  rpc VerifyMessage(VerifyMessageRequest) returns (VerifyMessageResponse);
+}
+
+// Request message for sign message with private key.
+message SignMessageWithPrivateKeyRequest {
+  // The message to sign.
+  string message = 1;
+  // The private key to sign message.
```
[the remark this answers is not in the document]
Comment (b00f) [reply www/grpc/proto/utils.proto:21]: Let move private key as first item. Important variables come first.

File: www/grpc/proto/utils.proto [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+message SignMessageWithPrivateKeyResponse {
+  // The signature of message.
+  string signature = 1;
+}
+
+// Request message verify message.
+message VerifyMessageRequest {
+  // The signed message.
+  string message = 1;
+  // The signature of message.
```
Comment (b00f) [inline www/grpc/proto/utils.proto:35]:
```suggestion
  // The signature of the message.
```

File: www/grpc/proto/utils.proto [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  string signature = 2;
+    // The message signer public key.
+  string public_key = 3;
+}
+
+// Response message containing the resualt of validation of signature and message.
+message VerifyMessageResponse {
+  bool is_valid = 1;
```
Comment (b00f) [inline www/grpc/proto/utils.proto:43]: // Indicates if the signature is valid (true) or not (false).

File: www/grpc/buf/grpc-gateway.config.yaml [hunk 14]
```diff
@@ -87,3 +87,13 @@ http:
...
+
+    - selector: pactus.Wallet.SignMessage
+      get: "/pactus/wallet/sign_message"
+
+    # Util APIs
+    - selector: pactus.Utils.SignMessageWithPrivateKey
+      get: "/pactus/Utils/Sign_Message_With_PrivateKey"
+
+    - selector: pactus.Utils.Verify_Message
+      get: "/pactus/Utils/Verify_Message"
```
Comment (Ja7ad) [inline www/grpc/buf/grpc-gateway.config.yaml:99]: Please don't write capital case in camel case.

```
verify_message
```

File: www/grpc/buf/grpc-gateway.config.yaml [hunk 15]
```diff
@@ -87,3 +87,13 @@ http:
 
     - selector: pactus.Wallet.GetTotalBalance
       get: "/pactus/wallet/get_total_balance"
+
+    - selector: pactus.Wallet.SignMessage
+      get: "/pactus/wallet/sign_message"
+
+    # Util APIs
+    - selector: pactus.Utils.SignMessageWithPrivateKey
+      get: "/pactus/Utils/Sign_Message_With_PrivateKey"
```
Comment (Ja7ad) [inline www/grpc/buf/grpc-gateway.config.yaml:96]: Please don't write capital case in camel case.

```
sign_message_with_privatekey
```

**Review by Ja7ad — CHANGES_REQUESTED** [review]

**Comment by Ja7ad** [comment]
@ZigBalthazar please register utils service in json-rpc server.

https://github.com/pactus-project/pactus/blob/main/www%2Fjsonrpc%2Fserver.go#L57-L60

**Comment by Ja7ad** [comment]
@ZigBalthazar Please generate proto for new changes.

**Review by b00f — APPROVED** [review]

**Comment by b00f** [comment]
@ZigBalthazar Thanks

**Review by Ja7ad — APPROVED** [review]

**Pull request merged**